core/src/main/java/hudson/ProxyConfiguration.java

```diff
@@ -259,6 +259,40 @@ public PasswordAuthentication getPasswordAuthentication() {
         return con;
     }
 
+    /** This method invokes {@link ProxyConfiguration#open(URL)} setting {@link HttpURLConnection#setInstanceFollowRedirects(boolean)} to true if the opened connection is http.
+     * In case of a redirection it will be follow it, <b>even if the redirection changes the protocol (from http to https or vice versa) using the value in Location header</b>.
+     * @since TODO
+     */
+    public static URLConnection openURLAndRedirect(URL url) throws IOException {
+        URLConnection con = ProxyConfiguration.open(url);
+        // A redirection from http to https (or vise versa) returns a 302 response status. Force redirection
+        int responseCode = HttpURLConnection.HTTP_OK;
+        // TODO responseCode always will be OK for non-http connections. Processing needed
+        if (con instanceof HttpURLConnection) {
+            // prevent problems from misbehaving plugins disabling redirects by default
+            HttpURLConnection httpCon = ((HttpURLConnection) con);
+            httpCon.setInstanceFollowRedirects(true);
+            responseCode = httpCon.getResponseCode();
+            if (HttpURLConnection.HTTP_OK != responseCode && HttpURLConnection.HTTP_MOVED_TEMP != responseCode && HttpURLConnection.HTTP_MOVED_PERM != responseCode && HttpURLConnection.HTTP_SEE_OTHER != responseCode) {
+                // No OK Status and no redirection
+                throw new IOException("Could not open " + url + ". Response code: " + responseCode);
+            } else if (HttpURLConnection.HTTP_OK != responseCode) {
+                // In case of redirection, we have to connect to the new URL
+                String redirection = con.getHeaderField("Location");
+                con = ProxyConfiguration.open(new URL(redirection));
```

Comment:
From a security standpoint, we should probably limit the allowed protocols to `https` and `http` here. I doubt anyone expects an `ftp` URL here.

Reply:
The original behaviour is not to avoid others protocols. Your suggestion is to add any kind of handling as suggested @oleg-nenashev in this [previous comment](https://github.com/jenkinsci/jenkins/pull/3128#discussion_r154620368), isn't it? Should it be part of the scope of this PR? I'm not sure about that.

In case of implementing this handling, as this call will be recursive, we have two options:
1. Check if the header field `Location` matches an http/https url and throwing an exception if it doesn't.
2. Call the method again, open the connection and if it's not an instance of HttpURLConnection throwing then the exception.
Which one would be the best one from the point of view of security?

Reply:
We could just do 1) and stop the loop, the result will be a document with a redirect we don't follow, and whoever uses it will end up with unexpected content.